torchprime/torch_xla_models/trainers/basic.py

```diff
@@ -0,0 +1,353 @@
+    model = apply_xla_patch_to_nn_linear(model)
+
+    # Annotate model weights and activations with sharding constraints to distribute
+    # the training across devices following the SPMD paradigm.
+    sharding_config = OmegaConf.to_container(self.config.model.sharding, resolve=True)
+    assert isinstance(sharding_config, dict), (
+      f"Sharding config {sharding_config} must be a dict"
+    )
+    model = shard_torch_xla_model_from_config(model, config=sharding_config)
+
+    # Rematerialize forward computation during the backward pass if requested.
+    model = self._add_checkpoint_offload_scan_model(model)
+    model = self._add_optimization_barrier_model(model)
+    self.model = model
+
+    # Set up optimizers
+    self.optimizer = Adafactor(
+      params=model.parameters(),
+      lr=self.config.optimizer.learning_rate,
+      relative_step=False,
+      scale_parameter=False,
+    )
+
+    # TODO: this OOMs the TPU.
+    # self._prime_optimizer()
+
+    self.lr_scheduler = get_scheduler(
+      name=self.config.lr_scheduler.type,
+      optimizer=self.optimizer,
+      num_warmup_steps=self.config.lr_scheduler.warmup_steps,
+      num_training_steps=self.config.max_steps,
+    )
+
+    # Execute all initialization work queued so far before starting training.
+    torch_xla.sync()
+
+  def _prime_optimizer(self):
+    for group in self.optimizer.param_groups:
+      for p in group["params"]:
+        p.grad = torch.zeros_like(p)
+        p.grad.requires_grad_(False)
+    self.optimizer.step()
+    torch_xla.sync()
+
+  def _get_train_dataloader(self):
+    if self.train_dataset is None:
+      raise ValueError("Trainer: training requires a train_dataset.")
+
+    num_replicas = xr.process_count()
+    logger.info(f"Num replicas: {num_replicas}")
+    if self.minibatch:
+      sampler = torch.utils.data.DistributedSampler(
+        self.train_dataset,
+        num_replicas=num_replicas,
+        rank=xr.process_index(),
+      )
+    else:
+      # Without minibatch, every process loads the global batch the same way.
+      sampler = torch.utils.data.DistributedSampler(
+        self.train_dataset,
+        num_replicas=1,
+        rank=0,
+      )
+    assert self.global_batch_size is not None
+    if self.minibatch:
+      # Each process loads the per-host batch size.
+      batch_size = self.global_batch_size // num_replicas
+    else:
+      # Each process will load the global batch, then discard the unneeded parts.
+      batch_size = self.global_batch_size
+    dataloader = DataLoader(
+      self.train_dataset,
+      # Data collator will default to DataCollatorWithPadding, so we change it.
+      collate_fn=default_data_collator,
+      batch_size=batch_size,
+      sampler=sampler,
+      drop_last=True,
+    )
+    loader = pl.MpDeviceLoader(
+      dataloader, self.device, input_sharding=self.input_sharding_spec
+    )
+    return loader
+
+  def _add_checkpoint_offload_scan_model(self, model: nn.Module):
```

Comment:
Do you expect to duplicate these functionality in the upcoming SFT trainer as well? Just trying to understand the intended structure.

We probably need sharding/scan/optimization-barrier/etc in SFT too.